apple_sdk_11_0: generate frameworks.nix

stephank opened the pull request:
###### Description of changes

This is an attempt to generate `frameworks.nix` using the Swift compiler. Via `swiftc -scan-dependencies`, the compiler outputs JSON we can easily consume. A script runs this on a very simple Swift module that does just `import SomeFramework`, for every framework in the input SDK.

This PR is currently split in four commits:

- The first commit adds just the script.
- The second commit replaces `frameworks.nix` with a generated version. Dependencies that would've been removed in this process were added back as manual overrides in `apple_sdk.nix`.
- The third commit fixes circular references in the Nix expression by removing some of the manual overrides.
- The fourth commit slightly changes the way we unpack the macOS SDK, so that we preserve all files. This is really just `SDKSettings.{plist,json}`, there wasn't much else missing, but those files happen to be essential for the Swift compiler to work with the SDK as downloaded by Nix.

This successfully builds stdenv. I also tested `swift`, in order to run the script itself. As an extra check, I also built `mpv`, which worked after a partial revert of #211864 (currently only on staging), and after adding an Accelerate framework dependency (but it appears mpv is similarly broken on master). Especially mpv also uses rust, gfortran, python and a whole bunch of things in its build closure, so I think that makes this decently tested?

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

stephank committed:
apple_sdk_11_0: script to generate frameworks.nix

stephank committed:
apple_sdk_11_0: generate frameworks.nix

stephank committed:
apple_sdk_11_0: fix circular framework references

Atemu commented on pkgs/os-specific/darwin/apple-sdk-11.0/apple_sdk.nix:
```diff
@@ -152,30 +152,86 @@ in rec {
+      TWAIN                   = { inherit Carbon; };
+      VideoDecodeAcceleration = { inherit CoreVideo; };
+      WebKit                  = { inherit ApplicationServices Carbon libobjc; };
+    };
```
These are quite a lot. How were these figured out?

Could this at least be partially automated somehow?

stephank replied:
These are the difference between the generated set and old set. So everything in the old `frameworks.nix` the generator could not find.

I think each of those requires some investigation to see if they are still needed. But for now I thought let's not remove anything to prevent breakage elsewhere.

stephank committed:
apple_sdk_11_0: simplify SDK unpacking

The SDK was missing SDKSettings files. This is usually not a problem for
Nix builds, because we generate our own fake SDK structure when
necessary (in xcbuild), but not having these files blocks using the
upstream Apple SDK in tooling such as gen-frameworks.py.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1958

viraptor commented on pkgs/os-specific/darwin/gen-frameworks.py:
```diff
@@ -0,0 +1,147 @@
+    ]
+    frameworks.sort()
+
+    # Determine the longest ame for padding output.
```
```suggestion
    # Determine the longest name for padding output.
```

viraptor reviewed (approved):
makes sense to me

Atemu reviewed (approved):
Diff LGTM

stephank commented:
Thanks for the reviews! I fixed the typo and rebased on staging, but no other changes.

viraptor commented:
I think it's good enough for staging. If there's a massive breakage, we can still spot it in Hydra and fix up/revert.
